### PR TITLE
Add io.github.nest_desktop.nest-desktop

### PR DESCRIPTION
**Pull request by babsey**
### Please confirm your submission meets all the criteria

- [x] Please describe the application briefly. 
Explore activity dynamics of spiking neuronal networks.
- [x] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [x] I have read and followed all the [Submission and licence requirements][reqs].
- [x] I have [built][build] and tested the submission locally.
- [x] I am an author/developer/upstream contributor to the project. If not, I contacted upstream developers about this submission. **Link:**

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

### CODE REVIEW

File: io.github.nest-desktop.yml [hunk 1]
```diff
@@ -0,0 +1,88 @@
+id: io.github.nest-desktop
```
Comment (bbhtt) [inline io.github.nest-desktop.yml:1]: did you see the application ID guidelines in PR checklist?

image

File: io.github.nest-desktop.yml [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+runtime: org.freedesktop.Platform
+runtime-version: "23.08"
+base: org.electronjs.Electron2.BaseApp
+base-version: "23.08"
```
Comment (bbhtt) [inline io.github.nest-desktop.yml:5]:
```suggestion
runtime-version: "24.08"
base: org.electronjs.Electron2.BaseApp
base-version: "24.08"
```

File: io.github.nest-desktop.yml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+command: nest-desktop-launch
+separate-locales: false
+rename-desktop-file: nest-desktop.desktop
+rename-icon: nest-desktop
+rename-appdata: nest-desktop
+finish-args:
+  # GPU acceleration if needed
+  - --device=dri
+  # Needs to save files locally
+  - --filesystem=home
```
Comment (bbhtt) [inline io.github.nest-desktop.yml:19]: does this work with filechooser portals?

File: io.github.nest-desktop.yml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  # XShm access
+  - --share=ipc
+  # Needs to talk to the network:
+  - --share=network
+  # Display access (X11 or Wayland)
+  # - --socket=fallback-x11
+  - --socket=system-bus
```
Comment (bbhtt) [inline io.github.nest-desktop.yml:26]: this is not allowed. you have to figure out the exact permissions needed.

Comment (bbhtt) [reply io.github.nest-desktop.yml:26]: https://docs.flatpak.org/en/latest/sandbox-permissions.html#d-bus-access

File: io.github.nest-desktop.yml [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+      - type: file
+        path: nest-desktop-launch
+      - type: file
+        path: nest-desktop.appdata.xml
+      - type: file
+        path: nest-desktop.desktop
+      - type: file
+        path: icons/nest-desktop-256x256.png
+      - type: file
+        path: icons/nest-desktop-512x512.png
```
Comment (bbhtt) [inline io.github.nest-desktop.yml:46]: these 3 should be in https://github.com/nest-desktop/nest-desktop

File: io.github.nest-desktop.yml [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        npm install yarn
+        rm package-lock.json
+      # Install node modules
+      - yarn install
+      # Build the electron app
+      - yarn app:build
+      # Bundle app and dependencies
+      - mv release/linux-unpacked /app/main
+      # Install app wrapper
+      - install -Dm755 -t ${FLATPAK_DEST}/bin/ ./run.sh
```
Comment (bbhtt) [inline io.github.nest-desktop.yml:68]:
```suggestion
```

File: nest-desktop-launch [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+# Start NEST Server as daemon
+export NEST_SERVER_DAEMON=1
+/app/bin/nest-server start
+
+# Start NESTML Server as daemon
+# export NESTML_SERVER_DAEMON=1
+# /app/bin/nestml-server start
+
+# Start NEST Desktop
+/app/bin/nest-desktop --no-sandbox
```
Comment (bbhtt) [inline nest-desktop-launch:22]: use zypak from the electron baseapp, drop no-sandbox https://docs.flatpak.org/en/latest/electron.html#launching-the-app

**Comment by babsey** [comment]
Thank you for the review. I will reply in each comment.

File: nest-desktop-launch (hunk 7 above)
Comment (babsey) [reply nest-desktop-launch:22]: image

Comment (babsey) [reply nest-desktop-launch:22]: It already contains `zypack-wrapper.sh` in `/app/bin/nest-desktop`.

Comment (bbhtt) [reply nest-desktop-launch:22]: then why are you passing no-sandbox?

File: io.github.nest-desktop.yml (hunk 5 above)
Comment (babsey) [reply io.github.nest-desktop.yml:46]: the last 3?

File: io.github.nest-desktop.yml (hunk 3 above)
Comment (babsey) [reply io.github.nest-desktop.yml:19]: Yes, it works with filechooser dialog.

File: io.github.nest-desktop.yml (hunk 1 above)
Comment (babsey) [reply io.github.nest-desktop.yml:1]: is `org.nest_simulator.nest-desktop` fine when the page exists?

File: io.github.nest-desktop.yml (hunk 5 above)
Comment (bbhtt) [reply io.github.nest-desktop.yml:46]: yes

image

Comment (babsey) [reply io.github.nest-desktop.yml:46]: I moved all files (icons, metainfo and desktop) to upstream repo (`https://github.com/nest-desktop/nest-desktop`)

File: io.github.nest-desktop.yml (hunk 1 above)
Comment (babsey) [reply io.github.nest-desktop.yml:1]: Would be `org.nest_simulator.nest-desktop` fine for the page `https://nest-simulator.org/nest-desktop`?

Comment (babsey) [reply io.github.nest-desktop.yml:1]: The domain belongs to the NEST community. NEST Desktop, a GUI for NEST Simulator, is my contribution to the community.

File: io.github.nest-desktop.yml (hunk 4 above)
Comment (babsey) [reply io.github.nest-desktop.yml]: I removed `--socket=system-bus`

File: io.github.nest-desktop.yml (hunk 1 above)
Comment (bbhtt) [reply io.github.nest-desktop.yml]: > The domain belongs to the NEST community. NEST Desktop, a GUI for NEST Simulator, is my contribution to the community.

Unless you control it you cannot use it. Use the `io.github.nest_desktop.nest-desktop`

File: io.github.nest-desktop.yml (hunk 4 above)
Comment (bbhtt) [reply io.github.nest-desktop.yml]: that warning is harmless, but you most likely need to figure out what permissions are needed with --log-session-bus/ --log-system-bus

File: io.github.nest-desktop.yml (hunk 3 above)
Comment (bbhtt) [reply io.github.nest-desktop.yml]: why is `home` still needed?

File: io.github.nest-desktop.yml (hunk 4 above)
Comment (babsey) [reply io.github.nest-desktop.yml]:
```
C1: -> org.freedesktop.DBus call org.freedesktop.DBus.Hello at /org/freedesktop/DBus
C2: -> org.freedesktop.DBus fake wildcarded AddMatch for org.freedesktop.portal
C3: -> org.freedesktop.DBus fake wildcarded AddMatch for org.nest_simulator.nest-desktop
C4: -> org.freedesktop.DBus fake wildcarded AddMatch for org.mpris.MediaPlayer2.org.nest_simulator.nest-desktop
C5: -> org.freedesktop.DBus fake ListNames
B1: <- org.freedesktop.DBus return from C1
B2: <- org.freedesktop.DBus signal org.freedesktop.DBus.NameAcquired at /org/freedesktop/DBus
B3: <- org.freedesktop.DBus return from C2
*SKIPPED*
B4: <- org.freedesktop.DBus return from C3
*SKIPPED*
B5: <- org.freedesktop.DBus return from C4
*SKIPPED*
B6: <- org.freedesktop.DBus return from C5
C6: -> org.freedesktop.DBus fake GetNameOwner for org.freedesktop.DBus
C7: -> org.freedesktop.DBus fake GetNameOwner for org.freedesktop.portal.Desktop
C8: -> org.freedesktop.DBus fake GetNameOwner for org.freedesktop.portal.Documents
C9: -> org.freedesktop.DBus fake GetNameOwner for org.freedesktop.portal.Flatpak
C10: -> org.freedesktop.DBus fake GetNameOwner for org.freedesktop.portal.IBus
*SKIPPED*
C11: -> org.freedesktop.portal.Flatpak call org.freedesktop.DBus.Properties.Get at /org/freedesktop/portal/Flatpak
B7: <- org.freedesktop.DBus return from C6
*SKIPPED*
B8: <- org.freedesktop.DBus return from C7
*SKIPPED*
B9: <- org.freedesktop.DBus return from C8
*SKIPPED*
B10: <- org.freedesktop.DBus return from C9
*SKIPPED*
B11: <- org.freedesktop.DBus return from C10
*SKIPPED*
B101: <- :1.4636 return from C11
C12: -> org.freedesktop.portal.Flatpak call org.freedesktop.DBus.Properties.Get at /org/freedesktop/portal/Flatpak
B102: <- :1.4636 return from C12
C1: -> org.freedesktop.DBus call org.freedesktop.DBus.Hello at /org/freedesktop/DBus
C2: -> org.freedesktop.DBus fake wildcarded AddMatch for org.freedesktop.portal
C3: -> org.freedesktop.DBus fake wildcarded AddMatch for org.nest_simulator.nest-desktop
C4: -> org.freedesktop.DBus fake wildcarded AddMatch for org.mpris.MediaPlayer2.org.nest_simulator.nest-desktop
C5: -> org.freedesktop.DBus fake ListNames
B1: <- org.freedesktop.DBus return from C1
B2: <- org.freedesktop.DBus signal org.freedesktop.DBus.NameAcquired at /org/freedesktop/DBus
B3: <- org.freedesktop.DBus return from C2
*SKIPPED*
B4: <- org.freedesktop.DBus return from C3
*SKIPPED*
B5: <- org.freedesktop.DBus return from C4
*SKIPPED*
B6: <- org.freedesktop.DBus return from C5
C6: -> org.freedesktop.DBus fake GetNameOwner for org.freedesktop.DBus
C7: -> org.freedesktop.DBus fake GetNameOwner for org.freedesktop.portal.Desktop
C8: -> org.freedesktop.DBus fake GetNameOwner for org.freedesktop.portal.Documents
C9: -> org.freedesktop.DBus fake GetNameOwner for org.freedesktop.portal.Flatpak
C10: -> org.freedesktop.DBus fake GetNameOwner for org.freedesktop.portal.IBus
*SKIPPED*
B7: <- org.freedesktop.DBus return from C6
*SKIPPED*
B8: <- org.freedesktop.DBus return from C7
*SKIPPED*
B9: <- org.freedesktop.DBus return from C8
*SKIPPED*
B10: <- org.freedesktop.DBus return from C9
*SKIPPED*
B11: <- org.freedesktop.DBus return from C10
*SKIPPED*
C11: -> org.freedesktop.DBus call org.freedesktop.DBus.AddMatch at /org/freedesktop/DBus
B12: <- org.freedesktop.DBus return from C11
C12: -> org.freedesktop.portal.Flatpak call org.freedesktop.portal.Flatpak.Spawn at /org/freedesktop/portal/Flatpak
B104: <- :1.4636 return from C12
B105: <- :1.4636 signal org.freedesktop.portal.Flatpak.SpawnStarted at /org/freedesktop/portal/Flatpak
[7:1105/152709.732011:ERROR:bus.cc(407)] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
```

File: io.github.nest-desktop.yml (hunk 1 above)
Comment (babsey) [reply io.github.nest-desktop.yml]: Ok, I am renaming it.

File: io.github.nest-desktop.yml (hunk 4 above)
Comment (babsey) [reply io.github.nest-desktop.yml]: Clicking on desktop file (on icon) works but not in terminal `flatpak run ..`.

File: io.github.nest-desktop.yml (hunk 3 above)
Comment (babsey) [reply io.github.nest-desktop.yml]: I am testing it without `home`.

Comment (babsey) [reply io.github.nest-desktop.yml]: Without `home` the filechooser doesn't work well. 

I need it to export data to json file or import data from json file.

File: io.github.nest-desktop.yml (hunk 4 above)
Comment (hfiguiere) [reply io.github.nest-desktop.yml]: to see the system bus you need to also add `--system-talk=something` (not sure if that value will pass, but any valid dbus name is enough) along the `--log-system-bus` otherwise it's closed

Comment (hfiguiere) [reply io.github.nest-desktop.yml]: there is no such name on the system bus, so no. The only reason this make the message disappear is that it exposes a filtered system bus. But if you don't know what name is needed, put none. It's just a bug.

Two wrong don't make a right, but you don't even copy the wrong properly at it says `'--talk-name=org.freedesktop.Notifications'` (and it's not necessary either)

File: flathub.json [hunk 8]
```diff
@@ -0,0 +1,5 @@
+{
+    "only-arches": [
+        "x86_64"
```
Comment (hfiguiere) [inline flathub.json:3]: this is disappointing.

File: io.github.nest_desktop.nest-desktop.yml [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  # Audio output
+  - --socket=pulseaudio
+  # Read/write home directory access
+  - --filesystem=home
+  # Chromium uses a socket in tmp for its singleton check
+  - --filesystem=/tmp
+  # Allow communication with network
+  - --share=network
+  # System notifications with libnotify
+  - --system-talk-name=org.freedesktop.Notifications
```
Comment (hfiguiere) [inline io.github.nest_desktop.nest-desktop.yml:30]:
```suggestion
```
this name doesn't exist on the system bus.

File: io.github.nest_desktop.nest-desktop.yml [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+  - --talk-name=org.freedesktop.Notifications
+modules:
+  - name: nest-desktop
+    buildsystem: simple
+    sources:
+      - type: git
+        url: https://github.com/nest-desktop/nest-desktop
+        branch: dev
```
Comment (hfiguiere) [inline io.github.nest_desktop.nest-desktop.yml:38]: NEVER a git branch

```suggestion
```

File: io.github.nest_desktop.nest-desktop.yml [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+      # Install app wrapper
+      - install -Dm0755 run.sh "${FLATPAK_DEST}/bin/nest-desktop"
+      - install -Dm0755 scripts/nest-desktop-launch "${FLATPAK_DEST}/bin/nest-desktop-launch"
+      - install -Dm644 icons/icon-256x256.png /app/share/icons/hicolor/256x256/apps/nest-desktop.png
+      - install -Dm644 icons/icon-512x512.png /app/share/icons/hicolor/512x512/apps/nest-desktop.png
+      - install -Dm644 io.github.nest_desktop.nest-desktop.metainfo.xml ${FLATPAK_DEST}/share/metainfo/nest-desktop.metainfo.xml
+      - install -Dm0755 nest-desktop.desktop "${FLATPAK_DEST}/share/applications/nest-desktop.desktop"
+
+  - name: nest-simulator
+    buildsystem: cmake
```
Comment (hfiguiere) [inline io.github.nest_desktop.nest-desktop.yml:73]:
```suggestion
    buildsystem: cmake-ninja
```

Unless there is a good reason not too

File: io.github.nest_desktop.nest-desktop.yml [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    sources:
+      - type: git
+        url: https://github.com/nest/nest-simulator
+        tag: v3.8
+        commit: 6dca9500e512da50fa730a7b62ce4b05c77e2bfd
+    build-options:
+      build-args:
+        - --share=network
```
Comment (hfiguiere) [inline io.github.nest_desktop.nest-desktop.yml:81]: like above, the build MUST be offline this is not allowed.

Comment (babsey) [reply io.github.nest_desktop.nest-desktop.yml:78]: Is it because of the security policy?

For both builds (nest-desktop and nest-simulator) it needs to install deps (`node_modules` and python packages).

What is the proper solution for such step?

File: flathub.json (hunk 8 above)
Comment (babsey) [reply flathub.json:3]: It is my first draft of flatpak building code. In general I have no experience if it is working for other architectures.

I need to discuss with NEST developer (https://github.com/nest/nest-simulator).

Comment (hfiguiere) [reply flathub.json:3]: isn't idea being Electron that it is all javascript? The arch shouldn't matter.

**Comment by babsey** [comment]
I apologize for delayed response.

Because the requirement to build it offline makes the current state impossible. 
Thus, I have to reconsider the concept.

I added modules for python packages with `flatpak_pip_generator`.
For node modules, I tried with `flatpak-node-generator`. It works when it only installs node modules from `yarn.lock`.
However it fails because electron needs to install its deps (in `postinstall` step). 

I decide to upload electron app to github repo and tell flatpak to download the file.

@hfiguiere @bbhtt I would like to ask you to review it again.

File: flathub.json (hunk 8 above)
Comment (babsey) [reply flathub.json:3]: That is correct, the frontend is all javascript. But the backend is `nest-simulator` running in cpp and it provides python interface. I asked NEST developer if `nest-simulator` is able to run in both architectures.

File: io.github.nest_desktop.nest-desktop.yml [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+  - --socket=x11
+  - --share=ipc
+  # Open GL
+  - --device=dri
+  # Audio output
+  - --socket=pulseaudio
+  # Read/write home directory access
+  - --filesystem=home
+  # Chromium uses a socket in tmp for its singleton check
+  - --filesystem=/tmp
```
Comment (bbhtt) [inline io.github.nest_desktop.nest-desktop.yml:22]: sandbox has its own /tmp why is this needed?

File: io.github.nest_desktop.nest-desktop.yml [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+  # Allow communication with network
+  - --share=network
+  # System notifications with libnotify
+  - --talk-name=org.freedesktop.Notifications
```
Comment (bbhtt) [inline io.github.nest_desktop.nest-desktop.yml:26]:
```suggestion
```

libnotify uses the portal by default now.

File: io.github.nest_desktop.nest-desktop.yml [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+modules:
+  - name: nest-desktop
+    buildsystem: simple
+    sources:
+      - type: git
+        url: https://github.com/nest-desktop/nest-desktop
+        commit: 6acecc1923605f2d4cb8d95c5ed49697b6fd1f98
+      - type: file
+        url: https://github.com/nest-desktop/nest-desktop-electron/releases/download/v4.0.0/nest-desktop-4.0.0.tar.gz
+        sha256: 9f83f4bed4ae74428518051e73b4c32e1090a04e46dc41d2187f6cf97b2d264b
```
Comment (bbhtt) [inline io.github.nest_desktop.nest-desktop.yml:36]:
```suggestion
        sha256: 9f83f4bed4ae74428518051e73b4c32e1090a04e46dc41d2187f6cf97b2d264b
        dest-filename: nest-desktop.tar.gz
```

File: io.github.nest_desktop.nest-desktop.yml [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+        dest-filename: run.sh
+        commands:
+          - exec zypak-wrapper.sh /app/main/nest-desktop "$@"
+      - type: file
+        path: io.github.nest_desktop.nest-desktop.metainfo.xml
+    build-options:
+      env:
+        APP_VERSION: 4.0.0
+    build-commands:
+      - tar -zxf nest-desktop-${APP_VERSION}.tar.gz -C ${FLATPAK_DEST}
```
Comment (bbhtt) [inline io.github.nest_desktop.nest-desktop.yml:47]:
```suggestion
    build-commands:
      - tar -zxf nest-desktop.tar.gz -C ${FLATPAK_DEST}
```

File: io.github.nest_desktop.nest-desktop.yml (hunk 12 above)
Comment (babsey) [reply io.github.nest_desktop.nest-desktop.yml]: For node modules, I tried with `flatpak-node-generator`. It works when it only installs node modules from yarn.lock. However it failed because electron needs to install its deps (in the `postinstall` step).

I decided to upload electron app to github repo and tell flatpak to download the file.

File: flathub.json (hunk 8 above)
Comment (babsey) [reply flathub.json]: They told me that it should work with `aarch64`. I removed `flathub.json`.

**Comment by bbhtt** [comment]
bot, build io.github.nest_desktop.nest-desktop

**Comment by babsey** [comment]
It failed because no metainfo in `appdata` found. I am confused, `appdata` is deprecated, isn't it?

```No metainfo file for io.github.nest_desktop.nest-desktop was found in /app/share/metainfo or /app/share/appdata```

File: io.github.nest_desktop.nest-desktop.yml [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+      - type: file
+        path: io.github.nest_desktop.nest-desktop.metainfo.xml
+    build-commands:
+      - tar -zxf nest-desktop.tar.gz -C ${FLATPAK_DEST}
+      - mv ${FLATPAK_DEST}/nest-desktop* ${FLATPAK_DEST}/main
+      - install -Dm0755 run.sh "${FLATPAK_DEST}/bin/nest-desktop"
+      - install -Dm0755 scripts/nest-desktop-launch "${FLATPAK_DEST}/bin/nest-desktop-launch"
+      - install -Dm644 icons/icon-256x256.png /app/share/icons/hicolor/256x256/apps/nest-desktop.png
+      - install -Dm644 icons/icon-512x512.png /app/share/icons/hicolor/512x512/apps/nest-desktop.png
+      - install -Dm644 io.github.nest_desktop.nest-desktop.metainfo.xml ${FLATPAK_DEST}/share/metainfo/nest-desktop.metainfo.xml
```
Comment (bbhtt) [inline io.github.nest_desktop.nest-desktop.yml:47]:
```suggestion
      - install -Dm644 io.github.nest_desktop.nest-desktop.metainfo.xml -t ${FLATPAK_DEST}/share/metainfo/
```

File: io.github.nest_desktop.nest-desktop.yml [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+runtime: org.freedesktop.Platform
+runtime-version: "24.08"
+base: org.electronjs.Electron2.BaseApp
+base-version: "24.08"
+sdk: org.freedesktop.Sdk
+command: nest-desktop-launch
+separate-locales: false
+rename-icon: nest-desktop
+rename-desktop-file: nest-desktop.desktop
+rename-appdata: nest-desktop
```
Comment (bbhtt) [inline io.github.nest_desktop.nest-desktop.yml:11]:
```suggestion
rename-appdata-file: nest-desktop.metainfo.xml
```

**Comment by bbhtt** [comment]
Metainfo file should also be moved upstream https://github.com/nest-desktop/nest-desktop/

See the note in https://docs.flathub.org/docs/for-app-authors/requirements#required-metadata

**Comment by babsey** [comment]
bot, build io.github.nest_desktop.nest-desktop

**Comment by babsey** [comment]
bot, build io.github.nest_desktop.nest-desktop

**Comment by babsey** [comment]
Unclear what causes the error in the built step.
```
ERROR: Could not find a version that satisfies the requirement poetry-core==1.9.1 (from poetry) (from versions: 1.9.0)
ERROR: No matching distribution found for poetry-core==1.9.1
```

**Comment by babsey** [comment]
bot, build io.github.nest_desktop.nest-desktop

**Comment by babsey** [comment]
It failed again because of unknown license in metainfo file. I fixed it and restart the build step.

**Comment by babsey** [comment]
bot, build io.github.nest_desktop.nest-desktop
